src/codegen.cpp

```diff
@@ -4327,7 +4379,12 @@ static jl_cgval_t emit_call_specfun_other(jl_codectx_t &ctx, bool is_opaque_clos
         setName(ctx.emission_context, TheCallee, namep);
     }
     CallInst *call = ctx.builder.CreateCall(cft, TheCallee, argvals);
-    call->setAttributes(returninfo.attrs);
+    if (auto fun = dyn_cast<Function>(TheCallee)){
+        fun->setAttributes(AttributeList::get(ctx.builder.getContext(), {get_attrs_ipoeffects(ctx.builder.getContext(), effects, true, has_ptr_arg), returninfo.attrs, attrs}));
```

Comment:
It is generally pretty sketchy to mutate the callee to assume information that was only valid for this particular caller call site. I would delete this change.

Reply:
The slightly annoying thing is that speculatable can't be passed to call sites, only on declarations. Also aren't the ipo effects the ones that are valid on call sites of the specific MI?